scipy/stats/stats.py

```diff
@@ -558,11 +558,12 @@ def tmean(a, limits=None, inclusive=(True, True), axis=None):
 
     """
     a = asarray(a)
+    a = a.astype(float)
     if limits is None:
-        return np.mean(a, None)
-
-    am = _mask_to_limits(a.ravel(), limits, inclusive)
-    return am.mean(axis=axis)
+        return np.mean(a, axis)
+    am = _mask_to_limits(a, limits, inclusive)
+    amnan = am.filled(fill_value=np.nan)
+    return np.nanmean(amnan, axis=axis)
```

Comment:
Are you sure the masked array mean is ignoring the `axis` argument. If so, that's a NumPy bug. Note that the example from the issue doesn't include limits, so never hits this code.

Reply:
Yes, although the original issue doesn't mention this, it leads to an `AxisError` on `master`:

```
>>> r = np.array([[1, 1, 1], [2, 2, 2], [3, 3, 3]])
>>> stats.tmean(r, limits=(1, 2), axis=1)

----> 1 stats.tmean(r, limits=(1,2), axis=1)
~/scipy/scipy/stats/stats.py in tmean(a, limits, inclusive, axis)                                                                                                                   563
    564     am = _mask_to_limits(a.ravel(), limits, inclusive)                                                                                                                  --> 565     return am.mean(axis=axis)
    566                                                                                                                                                                             567
<snip snip>
AxisError: axis 1 is out of bounds for array of dimension 1
```

The fix I introduced produces the correct answer: `array([ 1.,  2., nan])`. 

I'm pretty sure the culprit is [here](https://github.com/scipy/scipy/blob/8040f6706acf43f4cbdc4705fe507cd459a60fdc/scipy/stats/stats.py#L564), where the input array is raveled. The `axis` keyword becomes meaningless then.

Reply:
Yes, `ravel` would get rid of the axes. But you should be able to use `.mean(axis)` on the results without `ravel`.

Reply:
Yes, that will work, although it'll then return a `MaskedArray`. I did it this way for consistency with `scipy.stats.tvar`.

Reply:
There's already machinery to do this in `scipy.stats.mstats`. For example, `scipy.stats.mstats_basics.tmean` does the exact same thing:

```
>>> mstats.tmean(r, limits=(1,2), axis=1)
masked_array(data=[1.0, 2.0, --],
             mask=[False, False,  True],
       fill_value=1e+20)
```

I'm not sure, should we have two functions that do the exact same thing?

Reply:
Ah, I see. This is slightly weird in that it changes the treatment of input nans, but I suppose being consistent with `tvar` is still an improvement.

The other option would be to access the `.data` attribute on the masked array when an axis is passed in.

Reply:
Hmm, good point. How about something like this?
```suggestion
    am = _mask_to_limits(a, limits, inclusive)
    return am.mean(axis=axis).filled(fill_value=np.nan)
```

This propagates input nans to the output -- nans are only added and never disappear.

Reply:
That seems reasonable, although I don't think it works for `axis=None` where `mean` returns a scalar. You would need something like
```python
return am.mean() if axis is None else am.mean(axis).filled(fill_value=np.nan)
```

Reply:
Hrm, actually even that's not quite right because you can have `axis=0` for a 1d array and the result is still a scalar.

Reply:
I think this version is safe:
```python
mean = np.ma.filled(am.mean(axis=axis), fill_value=np.nan)
return mean if mean.ndim > 0 else mean.item()
```

Reply:
Thanks for the suggestions @peterbell10, I've incorporated then in the PR.